src/viur/shop/modules/discount.py

```diff
@@ -207,3 +209,80 @@ def current_automatically_discounts(self) -> list[SkeletonInstance]:
             discounts.append(skel)
         logger.debug(f'current {discounts=}')
         return discounts
+
+    def remove(
+        self,
+        discount_key: db.Key | None = None,
+    ) -> t.Any:
+
+        if not isinstance(discount_key, (db.Key, type(None))):
+            raise TypeError(f"discount_key must be an instance of db.Key")
+        cart_key = self.shop.cart.current_session_cart_key  # TODO: parameter?
+
+        discount_skel = self.viewSkel()
+
+        if not discount_skel.fromDB(discount_key):
+            raise errors.NotFound
+        try:
+            # Todo what we do when we have more than more condition
+            application_domain = discount_skel["condition"][0]["dest"]["application_domain"]
+        except KeyError:
+            raise InvalidStateError("application_domain not set")
+
+        if discount_skel["discount_type"] == DiscountType.FREE_ARTICLE:
+            for cart_skel in self.shop.cart.get_children(parent_cart_key=cart_key):
+                if cart_skel["discount"] and cart_skel["discount"]["dest"]["key"] == discount_skel["key"]:
+                    break
+            else:
+                raise errors.NotFound
+            self.shop.cart.cart_remove(
+                cart_key=cart_skel["key"]
+            )
+
+            return {  # TODO: what should be returned?
+                "discount_skel": discount_skel}
+        elif application_domain == ApplicationDomain.BASKET:
+            if discount_skel["discount_type"] in {DiscountType.PERCENTAGE, DiscountType.ABSOLUTE}:
+                self.shop.cart.cart_update(
+                    cart_key=cart_key,
+                    discount_key=None
+                )
+                return {  # TODO: what should be returned?
+                    "discount_skel": discount_skel,
+                }
+        elif application_domain == ApplicationDomain.ARTICLE:
+
+            applicable, dv = self.can_apply(discount_skel, cart_key)
```

Comment:
Why is it necessary to check whether a code is applicable if it is going to be removed anyway?

I think there's too much logic in this method anyway. Isn't it enough to simply run down the tree and remove all codes with the key?